rust/src/witness_builder.rs

```diff
@@ -0,0 +1,389 @@
+use std::collections::{HashSet, HashMap};
+use super::*;
+
+#[wasm_bindgen]
+#[derive(Clone, PartialEq, Eq,  Hash)]
+pub struct RedeemerWitnessKey {
+    tag: RedeemerTag,
+    index: BigNum,
+}
+
+#[wasm_bindgen]
+impl RedeemerWitnessKey {
+
+    pub fn tag(&self) -> RedeemerTag {
+        self.tag.clone()
+    }
+
+    pub fn index(&self) -> BigNum {
+        self.index.clone()
+    }
+
+    pub fn new(tag: &RedeemerTag, index: &BigNum) -> Self {
+        Self {
+            tag: tag.clone(),
+            index: index.clone(),
+        }
+    }
+
+}
+
+#[wasm_bindgen]
+#[derive(Clone)]
+pub struct RequiredWitnessSet {
+    vkeys: HashSet<Vkey>,
```

Comment:
I am assuming we will be constructing `RequiredWitnessSet` within the `TxBuilder`? If so, how are we supposed to get the `Vkey`? Wouldn't we only have the hash of it in addresses? Or are we gonna need to give it that info somehow like modifying `TxBuilderInput`? Or store the`Ed25519KeyHash` here instead? Although crypto calls are by far the slowest thing in this library so that could be a point of consideration.

Reply:
Hmmm. The actual key for uniqueness used in the witness data structure is the `VKey` but the mapping between `VKey` <-> `Ed25519KeyHash` should be unique in practice so you're right that we can deviate from the spec to make the interface more usable from the tx builder. I don't think this should cause any problems